### PR TITLE
Let MGXS be used in externally linked programs

### PR DESCRIPTION
**Pull request by gridley**
Hi all,

So, for my research, I can save a solid amount of work by leveraging code inside OpenMC. Unfortunately, there is much stuff that's tied to global variables and only works correctly when the proper sequence of OpenMC initialization routines have been called beforehand.

This PR (in my opinion) cleans up mgxs_interface a good bit, and removes the preponderance of global variables, leaving only a single MgxsInterface in OpenMC global scope. This is nice, because now, I can compile simple programs like this:

```cpp
#include "openmc/mgxs_interface.h"
#include "openmc/constants.h"

#include <iostream>
using namespace std;

int main()
{
  openmc::MgxsInterface mgxs("mgxs.h5", {"UO2"}, {{294.0}});

  // Now print some info about the first entry in the mgxs.h5 file:
  cout << mgxs.nuclides_MG.size() << endl;
  cout << "The first entry in mgxs.h5 is named " << mgxs.nuclides_MG[0].name << endl;
  cout << "Its total cross section in group 1 is "
       << mgxs.nuclides_MG[0].get_xs(openmc::MG_GET_XS_TOTAL, 1, nullptr, nullptr, nullptr)
       << endl;
}
```

Compiled like:

```
h5c++ test2.C -I/home/gavin/code/openmc/include -L/home/gavin/code/openmc/build/lib -lopenmc
```

And take advantage of previous work done in OpenMC, which can really save me some time. I hope the maintainers find this to be a useful contribution, and am happy to make as many changes to make this palatable for incorporation as are required.

Ideally, I'd like to keep adding some changes like this which allow OpenMC to be used as a general C++ library for doing Monte Carlo, which would be amazing for research. Opinions?

### CODE REVIEW

File: src/mgxs_interface.cpp [hunk 1]
```diff
@@ -19,41 +19,55 @@
...
+  read_header(path_cross_sections);
+  set_nuclides_to_read(xs_to_read);
+  set_nuclide_temperatures_to_read(xs_temps);
+  init();
+}
 
-//==============================================================================
-// Mgxs data loading interface methods
-//==============================================================================
+// Should these perhaps unnecessary setters be lumped into one?
```
[the remark this answers is not in the document]
Comment (gridley) [reply src/mgxs_interface.cpp]: Yeah, definitely makes sense to combine. No clue why I had done that...

File: src/mgxs_interface.cpp [hunk 2]
```diff
@@ -186,61 +182,114 @@ std::vector<std::vector<double>> get_mat_kTs()
...
-  file_close(file_id);
+  // Build vector of nuclide names which are to be read
+  std::vector<std::string> nuclide_names(data::nuclide_map.size());
+  for (const auto& kv : data::nuclide_map) {
+    nuclide_names[kv.second] = kv.first;
+  }
+
+  std::unordered_set<std::string> already_read;
+
+  // Loop over all files
```
[the remark this answers is not in the document]
Comment (gridley) [reply src/mgxs_interface.cpp]: yes i mind. Just kidding, yes, I have now changed this. :+1:

File: src/mgxs_interface.cpp [hunk 3]
```diff
@@ -186,61 +182,114 @@ std::vector<std::vector<double>> get_mat_kTs()
...
+  for (const auto& mat : model::materials) {
+    for (int i_nuc : mat->nuclide_) {
+      std::string& name = nuclide_names[i_nuc];
+
+      if (already_read.find(name) == already_read.end()) {
+        data::mgInterface.xs_to_read.push_back(name);
+        data::mgInterface.xs_temps_to_read.push_back(these_nuc_temps[i_nuc]);
+        // DBG
```
[the remark this answers is not in the document]
Comment (gridley) [reply src/mgxs_interface.cpp]: whoops, fixed, thanks for catching that!

**Comment by gridley** [comment]
Well, awesome, I'm glad to hear you like the changes! Your requested revisions should be in now.

**Comment by nelsonag** [comment]
Thanks @gridley; I'll wait the 36 hours to give anyone else a chance to weigh in before I merge.

File: include/openmc/mgxs_interface.h [hunk 4]
```diff
@@ -12,36 +12,71 @@
 namespace openmc {
 
 //==============================================================================
-// Global variables
+// Global MGXS data container structure
 //==============================================================================
 
-namespace data {
+struct MgxsInterface
+{
```
Comment (paulromano) [inline include/openmc/mgxs_interface.h]: Style guide says brace should go on line with struct

File: include/openmc/mgxs_interface.h [hunk 5]
```diff
@@ -12,36 +12,71 @@
...
+
+  std::vector<std::vector<double>> get_mat_kTs();
+
+  // Reads just the header of the cross sections file, to find
+  // min & max energies as well as the available XS
+  void read_header(const std::string& path_cross_sections);
+};
+
+namespace data {
+  extern MgxsInterface mgInterface;
```
Comment (paulromano) [inline include/openmc/mgxs_interface.h]: Variable names should be lowercase with underscores (see [naming](https://docs.openmc.org/en/latest/devguide/styleguide.html#naming) conventions)

File: include/openmc/mgxs_interface.h [hunk 6]
```diff
@@ -12,36 +12,71 @@
...
+  std::vector<Mgxs> macro_xs;
+
+  std::vector<double> energy_bins;
+  std::vector<double> energy_bin_avg;
+  std::vector<double> rev_energy_bins;
+
+  // temperatues of each available nuclide
+  std::vector<std::vector<double>> nuc_temps;
+
+  MgxsInterface() = default;
```
Comment (paulromano) [inline include/openmc/mgxs_interface.h]: We generally follow the C++ core guidelines, which suggest the [following order](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#nl16-use-a-conventional-class-member-declaration-order) for members in a class:
- type definitions
- constructors/destructors
- member functions
- data

File: include/openmc/mgxs_interface.h [hunk 7]
```diff
@@ -12,36 +12,71 @@
 namespace openmc {
 
 //==============================================================================
-// Global variables
+// Global MGXS data container structure
 //==============================================================================
 
-namespace data {
+struct MgxsInterface
```
Comment (paulromano) [inline include/openmc/mgxs_interface.h]: As @smharper discussed, usually `struct` implies the class is relatively simple. I would make this `class`

File: include/openmc/mgxs_interface.h (hunk 5 above)
Comment (paulromano) [reply include/openmc/mgxs_interface.h]: I think for simplicity, perhaps we should just call this `mg`. That makes accessing things out of it relatively simple, e.g.
```C++
data::mg.num_energy_bins
```
What do you think @nelsonag?

File: include/openmc/mgxs_interface.h [hunk 8]
```diff
@@ -12,36 +12,71 @@
...
+  std::vector<std::vector<double>> xs_temps_to_read;
 
-} // namespace data
+  // Name of the HDF5 file which contains mgxs
+  std::string cross_sections_path;
 
-//==============================================================================
-// Mgxs data loading interface methods
-//==============================================================================
+  std::vector<Mgxs> nuclides_MG;
```
Comment (paulromano) [inline include/openmc/mgxs_interface.h]: Now that this is guarded within a class, there's no reason to have the `_MG` suffix (which before helped prevent a name clash with `data::nuclides`)

File: src/mgxs_interface.cpp [hunk 9]
```diff
@@ -19,41 +19,52 @@
...
-std::vector<double> rev_energy_bins;
-
-} // namesapce data
+MgxsInterface::MgxsInterface(const std::string& path_cross_sections,
+                             const std::vector<std::string> xs_to_read,
+                             const std::vector<std::vector<double>> xs_temps)
+{
+  read_header(path_cross_sections);
+  set_nuclides_and_temperatures(xs_to_read, xs_temps);
+  init();
```
Comment (paulromano) [inline src/mgxs_interface.cpp]: I generally like to see `this->` when one member function is called from another just to make clear that it's not calling some other function, but I'll leave it up to you if you want to change it.

File: include/openmc/mgxs_interface.h [hunk 10]
```diff
@@ -12,36 +12,71 @@
...
+  std::vector<Mgxs> macro_xs;
+
+  std::vector<double> energy_bins;
+  std::vector<double> energy_bin_avg;
+  std::vector<double> rev_energy_bins;
```
Comment (paulromano) [inline include/openmc/mgxs_interface.h]: Data members in a class should have a [trailing underscore](https://docs.openmc.org/en/latest/devguide/styleguide.html#naming).

File: include/openmc/mgxs_interface.h (hunk 4 above)
Comment (gridley) [reply include/openmc/mgxs_interface.h]: ok, fixed

File: include/openmc/mgxs_interface.h (hunk 5 above)
Comment (gridley) [reply include/openmc/mgxs_interface.h]: I like the idea a lot, and have already made the change. So, if you don't approve @nelsonag, I will happily remove.

File: include/openmc/mgxs_interface.h (hunk 7 above)
Comment (gridley) [reply include/openmc/mgxs_interface.h]: Makes sense to me; I haven't thought about it that way.

File: include/openmc/mgxs_interface.h (hunk 8 above)
Comment (gridley) [reply include/openmc/mgxs_interface.h]: ah yes, true. That helps it look nice and clean to remove.

File: include/openmc/mgxs_interface.h (hunk 10 above)
Comment (gridley) [reply include/openmc/mgxs_interface.h]: Hm, OK, changed

File: src/mgxs_interface.cpp (hunk 9 above)
Comment (gridley) [reply src/mgxs_interface.cpp]: Hm, OK, not changed in that case.

File: include/openmc/mgxs_interface.h (hunk 5 above)
Comment (nelsonag) [reply include/openmc/mgxs_interface.h]: I agree with that naming

**Comment by nelsonag** [comment]
Thanks @gridley

**Pull request merged**